R/prob-brier.R

```diff
@@ -0,0 +1,167 @@
+#'
+#' @export
+brier_class <- function(data, ...) {
+  UseMethod("brier_class")
+}
+brier_class <- new_prob_metric(
+  brier_class,
+  direction = "minimize"
+)
+
+#' @export
+#' @rdname brier_class
+brier_class.data.frame <- function(data,
+                                   truth,
+                                   ...,
+                                   na_rm = TRUE,
+                                   case_weights = NULL) {
+  estimate <- dots_to_estimate(data, !!! enquos(...))
+
+  metric_summarizer(
+    metric_nm = "brier_class",
+    metric_fn = brier_class_vec,
+    data = data,
+    truth = !!enquo(truth),
+    estimate = !!estimate,
+    na_rm = na_rm,
+    case_weights = !!enquo(case_weights)
+  )
+}
+
+#' @rdname brier_class
+#' @export
+brier_class_vec <- function(truth,
+                            estimate,
+                            na_rm = TRUE,
+                            case_weights = NULL,
+                            ...) {
+  estimator <- finalize_estimator(truth, metric_class = "brier_class")
+
+  # estimate here is a matrix of class prob columns
+  brier_class_impl <- function(truth,
+                               estimate,
+                               ...,
+                               case_weights = NULL) {
+    check_dots_empty()
+
+    brier_class_estimator_impl(
+      truth = truth,
+      estimate = estimate,
+      case_weights = case_weights
+    )
+  }
+
+  metric_vec_template(
+    metric_impl = brier_class_impl,
+    truth = truth,
+    estimate = estimate,
+    na_rm = na_rm,
+    estimator = estimator,
+    case_weights = case_weights,
+    cls = c("factor", "numeric")
+  )
+}
+
+brier_class_estimator_impl <- function(truth,
+                                       estimate,
+                                       case_weights) {
+  brier_factor(
+    truth = truth,
+    estimate = estimate,
+    case_weights = case_weights
+  )
+}
+
+# If `truth` is already a vector or matrix of binary data
+brier_ind <- function(truth, estimate, case_weights = NULL) {
+  if (is.vector(truth)) {
+    truth <- matrix(truth, ncol = 1)
+  }
+
+  if (is.vector(estimate)) {
+    estimate <- matrix(estimate, ncol = 1)
+  }
+  # In the binary case:
+  if (ncol(estimate) == 1 & ncol(truth) == 2) {
+    estimate <- cbind(estimate, 1 - estimate)
+  }
+
+  resids <- (truth - estimate)^2
+
+  if (is.null(case_weights)) {
+    case_weights <- rep(1, nrow(resids))
+  }
+
+  not_missing <- stats::complete.cases(resids) & !is.na(case_weights)
+  resids <- resids[not_missing,, drop = FALSE]
+  case_weights <- case_weights[not_missing]
+
+  # Normalize weights (in case negative weights)
+  case_weights <- exp(case_weights) / sum(exp(case_weights))
+
+  res <- sum(resids * case_weights) / (2 * sum(case_weights) )
+  res
+}
+
+# When `truth` is a factor
+brier_factor <- function(truth, estimate, case_weights = NULL) {
+  if (!is.factor(truth)) {
+    rlang::abort("'truth' should be a factor.")
+  }
+  old_opt <- options()
```

Comment:
I don't particularly like this, wouldn't this leave us in a situation where a custom contrast function could theoretically change the behavior of this function?

Reply:
Yes. Should we depend on level hardboard and use that function?

Reply:
absolutely!